src/services/codefixes/convertTypedefToType.ts

```diff
@@ -0,0 +1,232 @@
+    })
+});
+
+function doChange(changes: textChanges.ChangeTracker, node: Node, sourceFile: SourceFile) {
+    if (containsTypeDefTag(node)) {
+        fixSingleTypeDef(changes, node, sourceFile);
+    }
+}
+
+function fixSingleTypeDef(
+    changes: textChanges.ChangeTracker,
+    typeDefNode: JSDocTypedefTag | undefined,
+    sourceFile: SourceFile,
+) {
+    if (!typeDefNode) return;
+
+    const declaration = createDeclaration(typeDefNode);
+    if(!declaration) return;
+
+    const comment = typeDefNode.parent;
+
+    changes.replaceNode(
+        sourceFile,
+        comment,
+        declaration
+    );
+}
+
+function createDeclaration(tag: JSDocTypedefTag): InterfaceDeclaration | TypeAliasDeclaration | undefined {
+    const { typeExpression } = tag;
+    if (!typeExpression) return;
+    const typeName = tag.name?.getFullText().trim();
+    if (!typeName) return;
+
+    // For use case @typedef {object}Foo @property{bar}number
+    // But object type can be nested, meaning the value in the k/v pair can be object itself
+    if (typeExpression.kind === SyntaxKind.JSDocTypeLiteral) {
+        return createInterfaceForTypeLiteral(typeName, typeExpression);
+    }
+    // for use case @typedef {(number|string|undefined)} Foo or @typedef {number|string|undefined} Foo
+    // In this case, we reach the leaf node of AST.
+    // Here typeExpression.type is a TypeNode, e.g. a UnionType or a primitive type, such as "number".
+    if (typeExpression.kind === SyntaxKind.JSDocTypeExpression) {
+     return createTypeAliasForTypeExpression(typeName, typeExpression);
+    }
+}
+
+function createInterfaceForTypeLiteral(
+    typeName: string,
+    typeLiteral: JSDocTypeLiteral
+    ): InterfaceDeclaration | undefined {
+    const propertySignatures = createSignatureFromTypeLiteral(typeLiteral);
+        if(!propertySignatures || propertySignatures.length === 0) return;
+        const interfaceDeclaration = factory.createInterfaceDeclaration(
+            [],
+            typeName,
+            [],
+            [],
+            propertySignatures,
+        );
+        return interfaceDeclaration;
+}
+
+function createTypeAliasForTypeExpression(
+    typeName: string,
+    typeExpression: JSDocTypeExpression
+    ): TypeAliasDeclaration | undefined {
+    const typeReference = createTypeReference(typeExpression.type);
+    if (!typeReference) return;
+    const declaration = factory.createTypeAliasDeclaration(
+        [],
+        factory.createIdentifier(typeName),
+        [],
+        typeReference
+    );
+    return declaration;
+}
+
+function createSignatureFromTypeLiteral(typeLiteral: JSDocTypeLiteral): PropertySignature[] | undefined {
+    const propertyTags = typeLiteral.jsDocPropertyTags;
+    if(!propertyTags || propertyTags.length === 0) return;
+    const props = (propertyTags as JSDocPropertyTag[]).reduce((
+        signatures: PropertySignature[], tag: JSDocPropertyTag
+        ) => {

```

Comment:
Generally, you want to use the `reduceLeft` helper instead, which is a lot faster.

Reply:
fixed